gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java

```diff
@@ -160,25 +163,40 @@ public void run()
       }
     } catch (InterruptedException interrupt) {
       log.info("Job interrupted by InterrupedException.");
-      interruptTaskExecution(countDownLatch);
+      interruptTaskExecution(Optional.of(countDownLatch));
     }
     log.info("All assigned tasks of job {} have completed in container {}", jobId, containerIdOptional.or(""));
   }
 
-  private void interruptTaskExecution(CountDownLatch countDownLatch)
-      throws InterruptedException {
+  /**
+   * A helper function that shutting down all outstanding tasks and destroy taskExecutor if timeout on waiting for
+   * certain tasks' termination.
+   */
+  private void interruptTaskExecution(Optional<CountDownLatch> countDownLatch) throws InterruptedException {
     log.info("Job interrupted. Attempting a graceful shutdown of the job.");
-    this.tasks.forEach(Task::shutdown);
-    if (!countDownLatch.await(5, TimeUnit.SECONDS)) {
-      log.warn("Graceful shutdown of job timed out. Killing all outstanding tasks.");
-      try {
+    this.shutdownTasks();
+    try {
+      if (countDownLatch.isPresent()) {
+        if (!countDownLatch.get().await(5, TimeUnit.SECONDS)) {
+          log.warn("Graceful shutdown of job timed out. Killing all outstanding tasks.");
+          this.taskExecutor.shutDown();
```

Comment:
Can this line be refactored outside the else {..} block to avoid it being called in both if and the else?